### PR TITLE
TT-6142 - AMQP support

### PR DESCRIPTION
**Pull request by bradleytrf**
https://truefoundry.atlassian.net/browse/TT-6142

### CODE REVIEW

File: async_processor/types.py [hunk 1]
```diff
@@ -288,6 +307,20 @@ def to_output(self) -> Output:
         return SQSOutput(self)
 
 
+class AMQPOutputConfig(OutputConfig):
+    type: constr(regex=r"^AMQP$") = "AMQP"
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:311]: Can we please make the casing consistent with all other configurations?

File: pyproject.toml [hunk 2]
```diff
@@ -26,6 +26,7 @@ cryptography = ">=41.0.6,<42.1.0"
 nats-py = { version = "2.6.0", optional = true }
 boto3 =  { version = ">=1.28.16,<1.29.0", optional = true }
 kafka-python = { version = ">=2.0.0,<2.1.0", optional = true }
+amqpstorm = {version = "^2.10.7", optional = true, extras = ["amqpstorm"]}
```
Comment (chiragjn) [inline pyproject.toml:29]: Just curious what was the decision process to pick this?
As in why not the more popular pika library? or celery's pyamqp?

File: async_processor/types.py [hunk 3]
```diff
@@ -216,6 +221,20 @@ def to_input(self) -> Input:
         return SQSInput(self)
 
 
+class AMQPInputConfig(InputConfig):
+    type: constr(regex=r"^amqp$") = "amqp"
+
+    hostname: str
+    port: Optional[int] = 5672
+    auth: AMQPUserPasswordAuth
```
Comment (debajyoti-truefoundry) [inline async_processor/types.py:229]: Should authentication always be mandatory?
Also, is this the same authentication method that https://console.cloud.google.com/marketplace/product/codes-public/cloudamqp.endpoints.codes-public.cloud.goog uses?

File: async_processor/amqp_pub_sub.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            **(config.auth.dict() if config.auth else {}),
+        ).channel()
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[str]]:
+        try:
+            response = await run_in_threadpool(
+                self._connection.basic.get, queue=self._queue_name
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:34]: For how long will this call block if there is nothing in the queue?

Comment (debajyoti-truefoundry) [reply async_processor/amqp_pub_sub.py:34]: Can this call throw a timeout error?

File: pyproject.toml (hunk 2 above)
Comment (debajyoti-truefoundry) [inline pyproject.toml:29]: Is there any library that is async already?

For example https://github.com/mosquito/aio-pika

Comment (debajyoti-truefoundry) [reply pyproject.toml:29]: Prefer async libraries if possible.

File: async_processor/amqp_pub_sub.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            port=self._port,
+            **(config.auth.dict() if config.auth else {}),
+        ).channel()
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[str]]:
+        try:
+            response = await run_in_threadpool(
```
Comment (debajyoti-truefoundry) [inline async_processor/amqp_pub_sub.py:33]: https://github.com/truefoundry/async_service/pull/66/files#r1524179011

File: pyproject.toml (hunk 2 above)
Comment (debajyoti-truefoundry) [reply pyproject.toml:29]: https://github.com/truefoundry/async_service/blob/main/async_processor/nats_pub_sub.py

**Pull request closed without merging**